# 🤔 Convolutional Neural Network 🤔

This is one way....

* It's possible to apply MLP (multi-layer perceptron) to image data
* Dataset we use for image is MNIST (handwritten digits)
* Dimensions of each image is 28x28
    * Can be repped by a **matrix** (2D tensor)
    * To use MLP, we **reshape** the image data into a single 1D tensor (1 column, **784** rows from 28x28)
    * And this is the input layer
* Loss we should choose is **categorical cross entropy**

<font size="+1">Multi-class --> softmax --> categorical cross-entropy</font>

<font size="+1">Multi-class --> multi-label --> sigmoid</font>

...AND NOW WE'RE GONNA FIND ANOTHER WAY! It mimics the human visual perception and understanding and translation of cognition

Enter **convolutional neural network**, which is architecturally *completely different* from MLP!

# 🧠 CNN

* **Super flipping good for images**
* CNN is basically 2D configuration of NN
* Input: **image** (3 NxN if *color*, NxN if *black-and-white*)
* Weights: **2D array**

<img src="rgb_image.png" width="300" height="300">

### Weights in CNN

* called **kernel** or **filter matrix**
* element-wise multiplication and addition to get the result
* You can have many kernels, which creates a **cube** of quasi-stacked matrices. This is a **hyperparameter** (and it *is possible to overfit*)

<img src="kernel_image.png" width="400" height="400">

### Stride

* the horizontal and vertical movement steps
* What is the output size with stride = 1 and stride = 2?
* ```output_size = (input_size - filter_size)/stride + 1```
    * Stride = 1: (7-3)/1 + 1 = 5
    * Stride = 2: (7-3)/2 + 1 = 3

<img src="stride_1.png" width="400" height="400">

<img src="stride_2.png" width="400" height="400">

* Stride visualization: http://deeplearning.stanford.edu/wiki/index.php/Feature_extraction_using_convolution

```python
model.add(Conv2D(32, kernel_size=(3, 3), strides=(1, 1), activation='relu', input_shape=input_shape))
```

### Max Pooling vs Average Pooling

* Max pooling: take the **maxiumum** element from each window of a certain size

<img src="maxpooling.png" width="400" height="400">

### Flatten Layer

* After feature extraction is completed by multiple *convolutional layers*, we use **flatten layer** to add MLP (*after convolutional layers*) to **classify**
* ```Keras```'s version of ```numpy.reshape```. Simple!
* Reshapes *n*-dimensional arrays to a vector -- necessary when moving from Conv2D layers, which expect 2-D arrays as inputs
    * Concrete example: A Flatten layer given a 28 x 28 array as input would output a vector of the shape (784, 1)

### Visualize the whole NN: CNN (feature learning) + MLP (classification)

<img src="CNN.png" width="600" height="600">

# 💚 Activity: Obtain the number of parameters for the following CNN

* By default, the strides = (1, 1)
* ```model.summary()``` tells us the **output shape** of our layers

In [2]:
from __future__ import print_function
import keras

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D


model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28, 28, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
# model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

print(model.summary())

Using TensorFlow backend.
/Users/cherishkim/anaconda3/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.6 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.7
  return f(*args, **kwds)


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1179776   
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
_________________________________________________________________


### Solutions

* 64 x 32 x 3 x 3 + 64 = **18496**
* 12 x 12 x 64 = **9216**
* 9216 x 128 + 128 = **1179776**
* 128 x 10 + 10 = **1290**
* ADD THEM ALL! = **1199882** --> That's the total # of parameters!

<img src="diagram.jpg" width="800">